mobile/examples/phi-3/ios/LocalLLM/LocalLLM/GenAIGenerator.mm

```diff
@@ -1,49 +1,143 @@
+        params->SetInputSequences(*sequences);
 
-  auto sequences = OgaSequences::Create();
-  tokenizer->Encode(prompt, *sequences);
+        auto generator = OgaGenerator::Create(*model, *params);
 
-  auto params = OgaGeneratorParams::Create(*model);
-  params->SetSearchOption("max_length", 200);
-  params->SetInputSequences(*sequences);
+        bool isFirstToken = true;
+        NSLog(@"Starting token generation loop...");
+        
+        startTime = Clock::now();
+        while (!generator->IsDone()) {
+            tokenStartTime = Clock::now();
 
-  // Streaming Output to generate token by token
-  auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
+            generator->ComputeLogits();
+            generator->GenerateNextToken();
 
-  auto generator = OgaGenerator::Create(*model, *params);
+            if (isFirstToken) {
+                firstTokenTime = Clock::now();
+                isFirstToken = false;
+            }
 
-  while (!generator->IsDone()) {
-    generator->ComputeLogits();
-    generator->GenerateNextToken();
+            // Get the sequence data and decode the token
+            const int32_t* seq = generator->GetSequenceData(0);
+            size_t seq_len = generator->GetSequenceCount(0);
+            const char* decode_tokens = tokenizer_stream->Decode(seq[seq_len - 1]);
 
-    const int32_t* seq = generator->GetSequenceData(0);
-    size_t seq_len = generator->GetSequenceCount(0);
-    const char* decode_tokens = tokenizer_stream->Decode(seq[seq_len - 1]);
+            if (!decode_tokens) {
+                @throw [NSException exceptionWithName:@"TokenDecodeError" reason:@"Token decoding failed." userInfo:nil];
+            }
 
-    NSLog(@"Decoded tokens: %s", decode_tokens);
+            // Measure token generation time excluding logging
+            TimePoint tokenEndTime = Clock::now();
+            auto tokenDuration = std::chrono::duration_cast<std::chrono::milliseconds>(tokenEndTime - tokenStartTime).count();
+            tokenTimes.push_back(tokenDuration);
+            NSString* decodedTokenString = [NSString stringWithUTF8String:decode_tokens];
+            [SharedTokenUpdater.shared addDecodedToken:decodedTokenString];
+        }
 
-    // Add decoded token to SharedTokenUpdater
-    NSString* decodedTokenString = [NSString stringWithUTF8String:decode_tokens];
-    [SharedTokenUpdater.shared addDecodedToken:decodedTokenString];
-  }
+        TimePoint endTime = Clock::now();
+        // Log token times
+        NSLog(@"Per-token generation times: %@", [self formatTokenTimes:tokenTimes]);
+
+        // Calculate metrics
+        auto totalDuration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+        auto firstTokenDuration = std::chrono::duration_cast<std::chrono::milliseconds>(firstTokenTime - startTime).count();
+
+        double promtProcTime =  (double)promptTokensCount / firstTokenDuration;
```

Comment:
should this be `promptTokens * 1000.0 / firstTokenDuration`?

and `promptProcRate` may be a better name.